File: DFS/46. Permutation/solution.py
https://leetcode.com/problems/permutations/
# Jiacheng Yang
# Runtime: 36 ms
# Memory Usage: 13.8 MB
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        visited = [0] * n
        
        def dfs(nums, visited, curr):
            if len(curr) == n:
                res.append(curr)
            for i in range(n):      # 每次i都是从0开始，因为可以往回找，visited来控制哪些找过
                if visited[i] == 0:
                    visited[i] = 1
                    dfs(nums, visited, curr + [nums[i]])
                    visited[i] = 0
        dfs(nums, visited, [])
        return res

File: DFS/46. Permutation/solution.cpp
https://leetcode.com/problems/permutations/
// Jiacheng Yang
// Runtime: 4 ms
// Memory Usage: 7.5 MB

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> visited(nums.size());
        vector<int> curr;
        dfs(nums, visited, curr, res);
        return res;            
    }
private:
    void dfs(const vector<int>& nums, vector<int>& visited, vector<int>& curr, vector<vector<int>>& res){
        if(curr.size() == nums.size())
            res.push_back(curr);
        for(int i = 0; i < nums.size(); i++)
        {
            if(!visited[i])
            {
                visited[i] = 1;
                curr.push_back(nums[i]);
                dfs(nums, visited, curr, res);
                curr.pop_back();
                visited[i] = 0;
            }
        }
    }
};

